scripts/filter_multimappers.py

```diff
@@ -133,13 +134,20 @@ def main(sam_file: Path) -> None:
         sam_file: Path to the input SAM file.
     """
     with pysam.AlignmentFile(sam_file, "r") as samfile:
+      
+        try:
+            first_alignment = next(samfile)
+            
+        except StopIteration:
+            sys.stdout.write(str(samfile.header))
+            return
 
         sys.stdout.write(str(samfile.header))
 
         current_query = None
         current_alignments: list[pysam.AlignedSegment] = []
 
-        for alignment in samfile:
+        for alignment in chain([first_alignment],samfile):
```

Comment:
If you _must_ extract the first alignment, couldn't you set `current_query` here, add the alignment to `current_alignments` and exclude `first_alignment` from the loop (no need for `chain()`. The you could also delete the `if current_query is None` check. Saves you two lines (three with the import), one pointless iteration and a few bytes of memory for the `chain()` function.

Reply:
By the way, I find the `write_output()` in the function below (I can't comment on that line, that's why I am writing it here) a bit opaque. At first sight, it seems that you are just reading and writing the file, without doing anything worthwhile. It is not obvious that the actual core of the script (the filtering of best alignments) is part of `write_output()`.

Why not write this in the main instead?

```py
with pysam.AlignmentFile(sam_file, "r") as samfile:

    ...

    for alignment in chain([first_alignment],samfile):

        ...

        else:
            if len(alignments) > 1:
                current_alignments = find_best_alignments(alns=current_alignments)
            write_output(alignments=current_alignments

    if len(alignments) > 1:
        current_alignments = find_best_alignments(alns=current_alignments)
    write_output(alignments=current_alignments)
```

(by the way, would be good to use either `alns` or `alignments` for both functions)

And then in `write_output()` just:

```py
for alignment in best_alignments:
    sys.stdout.write(alignment.to_string() + '\n')
```

And if you don't want to have the length check in `main()` (twice!), then just include it in `find_best_alignments()`. It totally makes sense to be in there anyway.